libsolidity/analysis/ControlFlowAnalyzer.cpp

```diff
@@ -191,6 +247,100 @@ void ControlFlowAnalyzer::checkUnreachable(CFGNode const* _entry, CFGNode const*
 		// Extend the location, as long as the next location overlaps (unreachable is sorted).
 		for (; it != unreachable.end() && it->start <= location.end; ++it)
 			location.end = std::max(location.end, it->end);
-		m_errorReporter.warning(5740_error, location, "Unreachable code.");
+
+		if (m_previousUnreachable.emplace(location).second)
+			m_errorReporter.warning(5740_error, location, "Unreachable code.");
 	}
 }
+
+bool ControlFlowAnalyzer::checkForReverts(ContractDefinition const& _contract, FunctionCall const& _functionCall)
+{
+	auto const& functionType = dynamic_cast<FunctionType const&>(
+		*_functionCall.expression().annotation().type
+	);
+
+	if (!functionType.hasDeclaration())
+		return false;
+
+	auto const& unresolvedFunctionDefinition =
+		dynamic_cast<FunctionDefinition const&>(functionType.declaration());
+
+	FunctionDefinition const* functionDefinition = nullptr;
+	ContractDefinition const* contractDefinition = nullptr;
+
+	SimpleASTVisitor visitor([&](ASTNode const& _node) {
+		if (auto const* memberAccess = dynamic_cast<MemberAccess const*>(&_node))
+		{
+			if (*memberAccess->annotation().requiredLookup == VirtualLookup::Super)
+			{
+				if (auto const typeType = dynamic_cast<TypeType const*>(memberAccess->expression().annotation().type))
+					if (auto const contractType = dynamic_cast<ContractType const*>(typeType->actualType()))
+					{
+						solAssert(contractType->isSuper(), "");
+						ContractDefinition const* superContract = contractType->contractDefinition().superContract(_contract);
+
+						functionDefinition =
+							&unresolvedFunctionDefinition.resolveVirtual(
+								_contract,
+								superContract
+							);
+					}
+			}
+			else
+			{
+				solAssert(*memberAccess->annotation().requiredLookup == VirtualLookup::Static, "");
+				functionDefinition = &unresolvedFunctionDefinition;
+			}
+			return false;
+		}
+		else if (auto const* identifier = dynamic_cast<Identifier const*>(&_node))
+		{
+			solAssert(*identifier->annotation().requiredLookup == VirtualLookup::Virtual, "");
+			functionDefinition = &unresolvedFunctionDefinition.resolveVirtual(_contract);
+
+			return false;
+		}
+		return true;
+	}, [](ASTNode const&){});
+
+	_functionCall.expression().accept(visitor);
+
+	solAssert(functionDefinition != nullptr, "");
+
+	if (!functionDefinition->isFree())
+		contractDefinition = functionDefinition->annotation().contract;
```

Comment:
Why are you using the contract the function is defined in?

Reply:
To find the correct functionFlow we need both the contract of the function and the function itself. As this is the function pointer AFTER the overload resolution, the contract it's defined in should always be the correct one.

Reply:
I think you still need to use `_contract` in this case. Even after resolving, the meaning of `super` is different in `functionDefinition` depending on the most derived contract.

Reply:
For the resolution we use the most derived contract. For the recursive checks, we use the current contract again. Only for finding the right functionflow we use this part here.

Reply:
So you are saying this is still correct because you do a new lookup every time you encounter a function call in the function flow? This sounds good, especially since it saves a lot of time.

But why do you need to store the functionFlow by contract then? Also `m_funcitonReverts[funcitonKey` - why does this work? A function might revert when called in one most derived contract but might not revert when called in another.

Reply:
> So you are saying this is still correct because you do a new lookup every time you encounter a function call in the function flow? This sounds good, especially since it saves a lot of time.

correct!

> But why do you need to store the functionFlow by contract then?

I think I was a bit confused when I wrote that part. I just went over the whole thing again and understand it properly now.
I don't need to store it by contract, indeed.

The `m_functionReverts` worked only because we have no test case with exactly that situation. I created one now:
```
contract A {
  function f() public virtual returns (uint) { g(); }
  function g() internal virtual { revert(); }
}
contract B is A {
  function f() public virtual override returns (uint) { A.f(); }
  function g() internal virtual override { A.g(); }
}
contract C is B {
  function f() public virtual override returns (uint) { A.f(); }
  function g() internal virtual override { }
}
// ----
// Warning 6321: (318-322): Unnamed return variable can remain unassigned. Add an explicit return with value to all non-reverting code paths or name the variable.

```
That's how it should look, the current code gives us another warning for `A.f()` at the moment. Currently fixing that.